applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/eigensolver_strategy.hpp

```diff
@@ -650,6 +660,45 @@ class EigensolverStrategy
         KRATOS_CATCH("")
     }
 
+    /// Normalize the eigenmodes with respect to the mass matrix
+    void MassNormalizeEigenvectors(DenseMatrixType& rEigenvectors)
+    {
+        const SparseMatrixType& rMassMatrix = this->GetMassMatrix();
+        const std::size_t num_modes = rEigenvectors.size1();
+        const std::size_t num_dofs  = rEigenvectors.size2();
+
+        DenseVectorType phi(num_dofs);
+        DenseVectorType tmp(num_dofs);
+
+        for (std::size_t j = 0; j < num_modes; ++j) {
+             // copy row j -> phi
+            IndexPartition<std::size_t>(num_dofs).for_each([&](std::size_t i){
+                phi[i] = rEigenvectors(j, i);
+            });
+
+            // tmp = M * phi_j
+            TSparseSpace::Mult(rMassMatrix, phi, tmp);
+
+            // Compute modal_mass = phi^T * tmp
+            const double modal_mass = IndexPartition<std::size_t>(num_dofs).for_each<SumReduction<double>>([&](std::size_t i)->double {
+                return phi[i] * tmp[i];
+            });
+
+            if (!(modal_mass > 0.0)) {
+                KRATOS_WARNING("MassNormalizeEigenvectors")
+                    << "Modal mass for mode " << j << " is non-positive (" << modal_mass << "). Skipping normalization.\n";
+                continue;
+            }
+
+            const double scale = 1.0 / std::sqrt(modal_mass);
+
+           // scale and write back into rEigenvectors 
+            IndexPartition<std::size_t>(num_dofs).for_each([&](std::size_t i){
```

Comment:
```suggestion
            IndexPartition<std::size_t>(num_dofs).for_each([&rEigenvectors, &phi, scale](std::size_t i){
```

I would only capture what is required only.

Reply:
Changed